### ISSUE TITLE
Faster way for annotation?

### ISSUE THREAD

**Issue by ritwikdubey** [AlexeyAB/darknet#1251]
Hi @AlexeyAB and other darknet enthusiasts,
I've used yolo_mark code for annotation so far. It's a great tool when the alternative is to manually create the text file but I don't think it's much of a scalable approach. I think there could be an alternative to manual tagging.

When the object detection is run on video stream, is it possible to save off those tagged video frames as images? My knowledge in OpenCV is quite shallow so I don't know if it's readily available feature or if it's something very simple. Thinking in layman terms, darknet ingests video stream frame by frame and performs object detection on one frame at a time. All one needs to do is save off those tagged frames. 

Someone might have asked this/done this/know this already, I'd appreciate if you could tell me how to do it. 

Ritwik

**Comment by ritwikdubey** [comment]
Well, I found a bunch of commented code lines in draw_detections_cv_v3() towards that effect. Any other bit of info I should know about it?

**Comment by AlexeyAB** [comment]
@ritwikdubey Hi, yes, 

1. you can train your model by using small number of labeld images to get `yolo-obj_8000.weights`

2. then do `./yolo_mark data/img cap_video test.mp4 10` to generate jpg-images (frames from some your video)

3. then do `./yolo_mark data/img data/train.txt data/obj.names` to generate `train.txt` - list of these images

4. and do pseudo labeling - to generated labels for these images `./darknet detector test data/obj.data yolo-obj.cfg yolo-obj_8000.weights -dont_show -save_labels < train.txt`
    more info: https://github.com/AlexeyAB/darknet/issues/717#issuecomment-390804310

5. then check correctnes of generated labels by using yolo_mark instead of manual labeling

6. then continue train your model by using this new `train.txt` that you should specify in `obj.data`
`./darknet detector train data/obj.data yolo-obj.cfg yolo-obj_8000.weights`

**Comment by ritwikdubey** [comment]
@AlexeyAB,
I tried something similar to your suggestion, but I wouldn't say I'm out of the woods just yet. Towards the effect of my initial question I did following:
1- Fetched a known/trained weight and config file for face recognition. Referred [here](https://github.com/dannyblueliu/YOLO-Face-detection)
2- Copied the code to save the label file to void `*detect_in_thread(void *ptr)` function.
```        
if (save_labels)
    {
        char* ext = ".txt";
        for (int c = 0; c < strlen(ext); ++c, ++i)
        {
            labelpath[i] = ext[c];
        }
        printf("%s \n", name);
        printf("%s \n", labelpath);

        FILE* fw = fopen(labelpath, "wb");
        //int i;
        for (i = 0; i < nboxes; ++i) {
            char buff[1024];
            int class_id = -1;
            float prob = 0;
            for (int j = 0; j < l.classes; ++j) {
                if (dets[i].prob[j] > demo_thresh && dets[i].prob[j] > prob) {
                    prob = dets[i].prob[j];
                    class_id = j;
                }
            }
            if (class_id >= 0) {
                sprintf(buff, "%d %2.4f %2.4f %2.4f %2.4f\n", class_id, dets[i].bbox.x, dets[i].bbox.y, dets[i].bbox.w, dets[i].bbox.h);
                fwrite(buff, sizeof(char), strlen(buff), fw);
            }
        }
        fclose(fw);
    }
```
I did it so that while running demo on live stream video the label file gets stored for every video frame.
3- Added code for saving video frame as well. 
```
    memcpy(predictions[demo_index], prediction, l.outputs*sizeof(float));
    mean_arrays(predictions, FRAMES, l.outputs, avg);
    l.output = avg;
     // save image file
    if (save_labels)
    {
        save_image_jpg(det_s, labelpath);
    }
    free_image(det_s);
```
4- `labelpath` is set so that both txt file image file gets dumped in same directory. 

Good so far. When I run demo I see the jpg and txt files get created.
```
C:\Users\rdub\Downloads\darknet\build\darknet\x64\data\obj\dump>ls
file_1.txt  file_1.jpg  file_2.txt  file_2.jpg  file_3.txt  file_3.jpg  file_4.txt file_4.jpg
```

Here's the problem. The label file shows entry like
```
0 249.953705 193.414719 96.070702 193.903595
```
which seems like it belongs to different scale compared to yolo_mark generated tag file. So I opened the corresponding jpg using yolo_mark and created a manual tag. The tag shows as

```
0 0.572266 0.480556 0.324219 0.569444
```

It's probably because of this that I see 70-90% detection of face in video stream but the bounding box doesn't get created. Maybe something done by yolo_mark is not done by the `save_labels` code. Any suggestion?

**Comment by ritwikdubey** [comment]
Rolling back a bit, I guess the problem is with cfg file/weights and not the code. I tried a known good weights file and cfg file. The bounding boxes, predictions and label files look similar to what I would expect.
```
34 0.7356 0.4480 0.0553 0.2914
4 0.6698 0.4521 0.0624 0.3184
7 0.3625 0.4687 0.0795 0.3589
27 0.8434 0.4370 0.0414 0.3040
26 0.7922 0.4418 0.0494 0.2853
27 0.5259 0.4615 0.0507 0.3405
26 0.4516 0.4631 0.0708 0.3289
0 0.5981 0.4581 0.0606 0.3111
33 0.9012 0.4310 0.0652 0.2902
12 0.2727 0.4716 0.0783 0.3628
```
![file_128](https://user-images.githubusercontent.com/1186630/43349817-2ab994c0-91c8-11e8-94e4-7dc06fe3291f.jpg)

**Comment by ritwikdubey** [comment]
I did not find an answer for the question, however, I looked up for more trained weights and cfg files for face detection. This error went away using the new files.

**Comment by ritwikdubey** [comment]
I did not find an answer for the question, however, I looked up for more trained weights and cfg files for face detection. This error went away using the new files.
